Integrated strain

gassmoeller opened the pull request:
This is a tracer property that tracks the integrated strain a certain piece of rock has experienced. I implemented it for illustration purposes, but it will be useful for strain-dependent rheologies as well.

bangerth commented:
We keep forgetting to ask, but this time I remember: please add a changelog entry. You'll also need to run the indentation script.

bangerth commented on source/particle/property/integrated_strain.cc:
```diff
@@ -0,0 +1,101 @@
+                                                              const std::vector<Tensor<1,dim> > &,
+                                                              std::vector<double> &data) const
+      {
+        for (unsigned int i = 0; i < Tensor<2,dim>::n_independent_components ; ++i)
```
How come you store the strain tensor as a full tensor, rather than a symmetric tensor?

gassmoeller replied:
This is analogous to the finite strain cookbook for compositional fields. I do not quite recall the reason (this central part is 6 months old now), but I think we had trouble multiplying a symmetric tensor from both sides with anti-symmetric tensors and store the result in a symmetric tensor, although the result should be symmetric again (@jdannberg, do you remember exactly why we decided to go for full tensors?)

bangerth replied:
I'd be curious to hear @jdannberg 's opinion.

My preference would be to store only the symmetric tensor and if we get into trouble somewhere multiplying, then we can do the conversion as appropriate at that location.

jdannberg replied:
I think in the beginning we just weren't sure if the strain tensor would remain symmetric, plus it was easier to implement for a full tensor. But it should be possible to use a symmetric tensor, I guess. For compositional fields that might even be more important than for the tracers (because we have to advect fewer fields). 
But like this, the implementations are the same for the tracers and compositional fields, which probably makes sense. If someone wants to use the finite strain in the material model, the same code can be used for both approaches. So if we change it here, we should probably change it for the fields too.

bangerth replied:
Yes, I think that would be my preference, to change it in both places. I'd also much rather do this now while it's still all new, than to train people to use the current way and later tell them to do it differently.

bangerth commented on source/particle/property/integrated_strain.cc:
```diff
@@ -0,0 +1,101 @@
+        for (unsigned int d=0; d<dim; ++d)
+          grad_u[d] = gradients[d];
+
+        const double dt = this->get_old_timestep();
```
Is this correct, or does it have to be the current time step? If correct, could you add a comment?

bangerth commented on source/particle/property/integrated_strain.cc:
```diff
@@ -0,0 +1,101 @@
+        const double dt = this->get_old_timestep();
+        const Tensor<2,dim> deformation_rate (symmetrize(grad_u));
+        const Tensor<2,dim> rotation (dt * (grad_u - deformation_rate) + unit_symmetric_tensor<dim>());
+        const Tensor<2,dim> new_strain = rotation * old_strain * transpose(rotation) + dt * deformation_rate;
```
Could you add a comment to this block that in formulas describes what you are doing here?

bangerth commented on source/particle/property/integrated_strain.cc:
```diff
@@ -0,0 +1,101 @@
+    {
+      ASPECT_REGISTER_PARTICLE_PROPERTY(IntegratedStrain,
+                                        "integrated strain",
+                                        "Implementation of a plugin in which the tracer "
```
Just say "A plugin..."

bangerth commented on source/particle/property/integrated_strain.cc:
```diff
@@ -0,0 +1,101 @@
+                                        "property is defined as the integrated strain this "
+                                        "particle has experienced. Note that the current "
+                                        "implementation assumes that in each timestep only "
+                                        "an infinite amount of strain is accumulated."
```
infinitesimal?

bangerth commented on source/particle/property/integrated_strain.cc:
```diff
@@ -0,0 +1,101 @@
+                                        "particle has experienced. Note that the current "
+                                        "implementation assumes that in each timestep only "
+                                        "an infinite amount of strain is accumulated."
+                                        "\n\n")
```
I don't think you need the two endlines.

bangerth commented:
Neat!

gassmoeller commented:
I think I addressed all comments, but it seems clang is not happy with it. The linker does not find `Tensor<2,dim>::n_independent_components`, but I have no clue why. @bangerth: Do you have an idea?

bangerth commented:
This looks like a bug in deal.II. Working on a patch.

bangerth commented:
By the way, I must be missing something, but I can't see the actual error message clang produces from the buildbot files on @tjhei 's machine. Do you have the exact error message?

bangerth commented on source/particle/property/integrated_strain.cc:
```diff
@@ -0,0 +1,116 @@
+
+        // the new strain is the rotated old strain plus the
+        // strain of the current time step
+        const Tensor<2,dim> new_strain = rotation * old_strain * transpose(rotation) + dt * deformation_rate;
```
This would be the place. If `old_strain` was just stored as a symmetric tensor, then you may have to do
```
  // convert old strain into a tensor so that it can be multiplied from the left and right by a non-symmetric tensor
  Tensor<2,dim> tmp = rotation * Tensor<2,dim>(old_strain) * transpose(rotation);
  // we know that the result must be a symmetric tensor, so convert that back and add this time step's deformation
  const Tensor<2,dim> new_strain = symmetrize(tmp) + deformation_rate * dt;
```

bangerth replied:
Oh, I meant this for the last line of course:
```
  const SymmetricTensor<2,dim> new_strain = symmetrize(tmp) + deformation_rate * dt;
```
You may want to convert `deformation_rate` into a `SymmetricTensor` as well.

bangerth commented:
By the way, I must be missing something, but I can't see the actual error message clang produces from the buildbot files on @tjhei 's machine. Do you have the exact error message?

bangerth commented:
@gassmoeller -- do you have clang installed on your machine? If so, can you see whether this patch helps: https://github.com/dealii/dealii/pull/2667

gassmoeller commented:
dealii/dealii#2667 does not seem to help on my machine (dealii and aspect compiled with clang 3.4.0). I reran cmake to catch the new files, but still the linker fails with:

CMakeFiles/aspect.dir/source/particle/property/integrated_strain.cc.o: In function `aspect::Particle::Property::IntegratedStrain<2>::get_property_information() const':
/home/bob/source/include/aspect/particle/property/integrated_strain.h:91: undefined reference to `dealii::Tensor<2, 2, double>::n_independent_components'
CMakeFiles/aspect.dir/source/particle/property/integrated_strain.cc.o: In function `aspect::Particle::Property::IntegratedStrain<3>::get_property_information() const':
/home/bob/source/include/aspect/particle/property/integrated_strain.h:91: undefined reference to `dealii::Tensor<2, 3, double>::n_independent_components'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
build FAILED

I found a workaround in my code though. If I store `n_independent_components` in a `const unsigned int` and then use that one in `std::make_pair` in integrated_strain.cc::92 the linker succeeds. I would go with that solution for now, so we do not have a new dependency on deal.II 8.5.0.pre. Would be nice to find the problem in deal though.

bangerth commented:
I updated the deal.II patch to actually do what I intended to do. I'd be interested if you could try it again.

But either way, if your approach works, then that's good enough. Let's just go with that then.

gassmoeller commented:
dealii/dealii#2667 has solved the problem for the Tensor class. I now modified the implementation to use SymmetricTensor, but of course the same problem appears there. I will take a look, but it seems more complicated to redo dealii/dealii#2667, because `SymmetricTensor` also defines internal Accessors and StorageTypes that need to be instantiated. Will take a look later.

bangerth commented:
Oh, I see :-(

But I think that shouldn't actually be a problem. Both classes are entirely in header files. It is not strictly necessary to instantiate these classes, all that needed to happen was to instantiate the static const members. That was just one part of the patch. I think you should essentially be able to take the entire patch, replace `Tensor` by `SymmetricTensor`, and that should work. No need to instantiate accessors and storage types (unless they also have static const members, but I think you could also ignore this for now).

bangerth commented:
There are currently two tests that fail. Can you take a look?

I've lost track whether this is otherwise ready to merge. Let me know if you think it's ready!

gassmoeller commented:
I updated the test results and resolved conflicts with master. When the tester is finished and happy, this should be ready to merge.

bangerth commented:
https://github.com/dealii/dealii/pull/2700 fixes the issue with the missing variable for `SymmetricTensor`.

bangerth commented on source/particle/property/integrated_strain.cc:
```diff
@@ -0,0 +1,117 @@
+                                        "defined as the integrated strain this "
+                                        "particle has experienced. Note that the current "
+                                        "implementation assumes that in each timestep only "
+                                        "an infinitesimal amount of strain is accumulated."
```
?? It is definitely a finite amount you add up. I think the sentence is more confusing then helpful -- please just remove.

bangerth commented on source/particle/property/integrated_strain.cc:
```diff
@@ -0,0 +1,117 @@
+                                        "particle has experienced. Note that the current "
+                                        "implementation assumes that in each timestep only "
+                                        "an infinitesimal amount of strain is accumulated."
+                                        "\n")
```
no need

bangerth commented:
OK with the minor changes.

gassmoeller commented:
Ok, done. Thanks for reviewing. I will merge when the tester is finished.
